engine/runtime-parser/src/main/java/org/enso/compiler/core/Persistance.java

```diff
@@ -0,0 +1,534 @@
+package org.enso.compiler.core;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataInput;
+import java.io.DataInputStream;
+import java.io.DataOutput;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.io.OutputStream;
+import java.nio.ByteBuffer;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.Map;
+import java.util.ServiceLoader;
+
+import org.enso.compiler.core.Persistance.Reference;
+
+public abstract class Persistance<T> {
+  private final Class<T> clazz;
+  private final boolean includingSubclasses;
+  private final int id;
+
+  protected Persistance(Class<T> clazz, boolean includingSubclasses, int id) {
+    this.clazz = clazz;
+    this.includingSubclasses = includingSubclasses;
+    this.id = id;
+  }
+
+  protected abstract void writeObject(T obj, Output out) throws IOException;
+  protected abstract T readObject(Input in) throws IOException, ClassNotFoundException;
+
+  @Override
+  public String toString() {
+    StringBuilder sb = new StringBuilder();
+    sb.append("Persistance{");
+    sb.append("clazz=").append(clazz.getName());
+    sb.append(", id=").append(id);
+    sb.append('}');
+    return sb.toString();
+  }
+
+  public static interface Output extends DataOutput {
+    public abstract <T> void writeInline(Class<T> clazz, T obj) throws IOException;
+    public abstract void writeObject(Object obj) throws IOException;
+  }
+
+  public static interface Input extends DataInput {
+    public abstract <T> T readInline(Class<T> clazz) throws IOException;
+    public abstract Object readObject() throws IOException;
+    public <T> Reference<T> readReference(Class<T> clazz) throws IOException;
```

Comment:
Loads a __lazy reference__ to an object of given `clazz`. The object isn't reconstructed immediately, but only when requested by calling `Reference.get(clazz)`. Allows creation of lazy `Seq` and other structures.